File: backend_node/src/controllers/usuarios.controller.js
const { getConnection, sql } = require("../config/db");
const bcrypt = require("bcrypt");
const jwt = require("jsonwebtoken");

// MÓDULO 1: REGISTRO DE USUARIOS
const registrarUsuario = async (req, res) => {
  try {
    const { username, email, password, confirmPassword, profile_image_url } =
      req.body;

    // 1. Validar que las contraseñas coincidan
    if (password !== confirmPassword) {
      return res.status(400).json({ mensaje: "Las contraseñas no coinciden" });
    }

    const pool = await getConnection();

    // 2. Validar que el usuario o correo no existan ya
    const checkUser = await pool
      .request()
      .input("username", sql.VarChar, username)
      .input("email", sql.VarChar, email)
      .query(
        "SELECT uid FROM users WHERE username = @username OR email = @email",
      );

    if (checkUser.recordset.length > 0) {
      return res
        .status(400)
        .json({ mensaje: "El nombre de usuario o correo ya está en uso" });
    }

    // 3. Encriptar la contraseña (10 rondas de salt)
    const hashedPassword = await bcrypt.hash(password, 10);

    // 4. Insertar en la base de datos
    await pool
      .request()
      .input("username", sql.VarChar, username)
      .input("email", sql.VarChar, email)
      .input("password", sql.VarChar, hashedPassword)
      .input("profile_image_url", sql.VarChar, profile_image_url).query(`
                INSERT INTO users (username, email, password, profile_image_url)
                VALUES (@username, @email, @password, @profile_image_url)
            `);

    res.status(201).json({ mensaje: "Usuario registrado exitosamente" });
  } catch (error) {
    console.error("Error en registro:", error);
    res.status(500).json({ mensaje: "Error interno del servidor" });
  }
};

// MÓDULO 2: INICIO DE SESIÓN (LOGIN)
const loginUsuario = async (req, res) => {
  try {
    const { username, password } = req.body;

    const pool = await getConnection();

    // 1. Buscar al usuario por su username
    const result = await pool
      .request()
      .input("username", sql.VarChar, username)
      .query("SELECT * FROM users WHERE username = @username");

    const user = result.recordset[0];

    if (!user) {
      return res.status(401).json({ mensaje: "Credenciales inválidas" });
    }

    // 2. Comparar la contraseña ingresada con el hash de la BD
    const isMatch = await bcrypt.compare(password, user.password);

    if (!isMatch) {
      return res.status(401).json({ mensaje: "Credenciales inválidas" });
    }

    // 3. Generar el Token JWT con el UID del usuario
    // El token durará 24 horas
    const token = jwt.sign(
      { uid: user.uid, username: user.username },
      process.env.JWT_SECRET,
      { expiresIn: "24h" },
    );

    // 4. Devolver el token y datos básicos (sin la contraseña)
    res.status(200).json({
      mensaje: "Login exitoso",
      token: token,
      usuario: {
        uid: user.uid,
        username: user.username,
        profile_image_url: user.profile_image_url,
      },
    });
  } catch (error) {
    console.error("Error en login:", error);
    res.status(500).json({ mensaje: "Error interno del servidor" });
  }
};

module.exports = {
  registrarUsuario,
  loginUsuario,
};

File: backend_python/src/controllers/usuarios_controller.py
import bcrypt
import jwt
import os
import datetime
from fastapi import HTTPException
from src.config.db import get_connection

JWT_SECRET = os.getenv("JWT_SECRET")

def registrar_usuario(data):
    if data.password != data.confirmPassword:
        raise HTTPException(status_code=400, detail="Las contraseñas no coinciden")

    conn = get_connection()
    cursor = conn.cursor()

    try:
        # 1. Validar que no exista
        cursor.execute(
            "SELECT uid FROM users WHERE username = %s OR email = %s", 
            (data.username, data.email)
        )
        if cursor.fetchone():
            raise HTTPException(status_code=400, detail="El nombre de usuario o correo ya está en uso")

        # 2. Encriptar contraseña
        hashed_password = bcrypt.hashpw(data.password.encode('utf-8'), bcrypt.gensalt())

        # 3. Insertar
        cursor.execute("""
            INSERT INTO users (username, email, password, profile_image_url)
            VALUES (%s, %s, %s, %s)
        """, (data.username, data.email, hashed_password.decode('utf-8'), data.profile_image_url))
        
        conn.commit()
        return {"mensaje": "Usuario registrado exitosamente"}

    except Exception as e:
        print(f"Error registro: {e}")
        # Re-lanzar HTTPExceptions, capturar el resto
        if isinstance(e, HTTPException):
            raise e
        raise HTTPException(status_code=500, detail="Error interno del servidor")
    finally:
        conn.close()

def login_usuario(data):
    conn = get_connection()
    cursor = conn.cursor()

    try:
        cursor.execute("SELECT * FROM users WHERE username = %s", (data.username,))
        user = cursor.fetchone()

        if not user:
            raise HTTPException(status_code=401, detail="Credenciales inválidas")

        # Comparar contraseñas
        if not bcrypt.checkpw(data.password.encode('utf-8'), user['password'].encode('utf-8')):
            raise HTTPException(status_code=401, detail="Credenciales inválidas")

        # Generar Token JWT (Expira en 24h)
        payload = {
            "uid": user['uid'],
            "username": user['username'],
            "exp": datetime.datetime.utcnow() + datetime.timedelta(hours=24)
        }
        token = jwt.encode(payload, JWT_SECRET, algorithm="HS256")

        return {
            "mensaje": "Login exitoso",
            "token": token,
            "usuario": {
                "uid": user['uid'],
                "username": user['username'],
                "profile_image_url": user['profile_image_url']
            }
        }

    except Exception as e:
        print(f"Error login: {e}")
        if isinstance(e, HTTPException):
            raise e
        raise HTTPException(status_code=500, detail="Error interno del servidor")
    finally:
        conn.close()